Documentation for ros2_control (#36)
ros-controls/roadmap: design_drafts/ros2_control_documentation.rst

When executing the control-loop ``update()`` method is called.
The method can access the latest hardware states and enables the controller to write the hardware's command interfaces.

User Interfaces
---------------
Users interact with the ros2_control framework using `Controller Manager`_'s services.
Those can be used directly or through the "cli"-interface (base command: ``ros2 control``).
The "cli"-interface is more user-friendly for direct interaction, i.e., outside of a node.

For a list of services and their definitions, check the ``srv`` folder in the `controller_manager_msgs`_ package.

For the description of the "cli"-interface, see the ``README.md`` file of the `ros2controlcli`_ package.

Hardware Components
===================
The *hardware components* realize communication to physical hardware and represent its abstraction in the ros2_control framework.
The components have to be exported as plugins using ``pluginlib``-library.
The `Resource Manager`_ dynamically loads those plugins and manages their lifecycle.

There are three basic types of components:

> Actuators

Reviewer: It would be more intuitive for a first reader to have the order flipped here: system, sensors, actuators --- as this is the order of importance for the majority of users

Author: You are right, I was just listed it alphabetically...

  
System
  Complex (multi-DOF) robotic hardware like industrial robots.
  The main difference between the *Actuator* component is the possibility to use complex transmissions like needed for humanoid robot's hands.
  This component has reading and writing capabilities.
  It is used when the is only one logical communication channel to the hardware (e.g., KUKA-RSI).  

A detailed explanation of hardware components is given in the `Hardware Access through Controllers design document`_.

Hardware Description in URDF
----------------------------
The ros2_control framework uses the ``<ros2_control>``-tag in the robot's URDF file to describe its components, i.e., the hardware setup.
The chosen structure enables tracking together multiple `xacro`-macros into one without any changes. 
The example hereunder shows a position-controlled robot with 2-DOF (RRBot), an external 1-DOF force-torque sensor, and an externally controlled 1-DOF parallel gripper as its end-effector.
For more examples and detailed explanations, check `ros-controls/ros2_control_demos`_ repository and `ROS2 Control Components URDF Examples design document`_.

.. code:: xml

> <ros2_control name="RRBotSystemPositionOnly" type="system">

Reviewer: the "rich view" mode doesn't seem to display this section properly

Reviewer 2: Afaik all lines which should be part of the block need to be indented.

Author: @gavanderhoorn thanks that did it!

#. Create a launch file to start the node with `Controller Manager`_.
   You can use a default `ros2_control node`_ (recommended) or integrate the controller manager in your software stack.
   (`Example launch file for RRBot <https://github.com/ros-controls/ros2_control_demos/blob/master/ros2_control_demo_robot/launch/rrbot_system_position_only.launch.py>`_)
   
Repositories
============
The ros2_control framework consist of the following repositories:

ros2_control
  The `ros2_control`_ repository implements the main interfaces and components of the framework mentioned in the previous sections.
  
ros2_controllers
  The `ros2_controllers`_ repository implements widely used controllers, e.g., forward controller, joint trajectory controller, differential drive controller, etc.
  
ros2_control_demos
  The `ros2_control_demos`_ repository provides examples of using the framework and templates for a smooth start with it.

Differences to ros_control (ROS1)
=================================

Hardware Structures - classes
-----------------------------

> The ros_control uses the ``RobotHW`` class as a rigid structure to handle any hardware.

Reviewer: ```suggestion
The ros_control framework uses the ``RobotHW`` class as a rigid structure to handle any hardware.
```

> This makes it impossible to extend the existing robot with additional hardware, like sensors, actuators, and tools, without coding.

Reviewer: `CombinedRobotHardware` is actually quite widely used, although an opt-in feature, not supported by default.

> The ros2_control defines three different types of hardware ``Actuator``, ``Sensor`` and ``System``.

Reviewer: ```suggestion
The ros2_control framework defines three different types of hardware ``Actuator``, ``Sensor`` and ``System``.
```

The ros2_control does not mandate a fixed set of interface types, but they are defined as strings in `hardware's description <#hardware-description-in-urdf>`__.
To ensure compatibility of standard controllers, standard interfaces are defined as constants in `hardware_interface package <https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/include/hardware_interface/types/hardware_interface_type_values.hpp>`__.

Controller's Access to Hardware
-------------------------------

In ros_control, the controllers had direct access to the ``RobotHW`` class requesting access to its interfaces (joints).
The hardware itself then took care of registered interfaces and resource conflicts.

In ros2_control, ``ResourceManager`` takes care of the state of available interfaces in the framework and enables controllers to access the hardware.
Also, the controllers do not have direct access to hardware anymore, but they register their interfaces to the `ControllerManager`.

Migration Guide to ros2_control
===============================

RobotHardware to Components
---------------------------

> #. Forget your implementation or the ``RobotHW`` interface. This is not used anymore. (Do not delete it, you can still extract some code.)

Reviewer: This is a little harsh :D I'd say that typically, one's `RobotHW` can be migrated directly to a `SystemInterface`, if one wants more granularity, they can branch some parts out into `SensorInterface` or `ActuatorInterface`

#. Decide which component type is suitable for your case. Maybe it makes sense to separate ``RobotHW`` into multiple components.
#. Implement `ActuatorInterface`_, `SensorInterface`_ or `SystemInterface`_ classes as follows:
   
   #. In the constructor, initialize all variables needed for communication with your hardware or define the default one.

>    #. In the configure function, read all the parameters your hardware needs from the parsed URDF snippet (i.e., from the `HardwareInfo`_ structure). Here you can cross-check if all joints and interfaces in URDF have allowed values or a combination of values.

Reviewer: these components can do their own parsing and are not limited to HardwareInfo

Author: Is it so? I am not sure what you mean. We provide all the data in the `HardwareInfo` structure. so you are actually not parsing, but iterating through the parsed structure.

Reviewer: You are right, I didn't remember precisely. This however is actually a good opportunity to lay out an example URDF that has a couple of custom parameters and we could show how these are appearing for a SystemInterface implementation. Perhaps a next ticket? I'd be happy to merge this as-is

      This usually includes changing the hardware state to receive commands or set it into a safe state before interrupting the command stream. 
      It can also include starting and stopping communication.
   #. Implement `read` and `write` methods to exchange commands with the hardware.
      This method is equivalent to those from `ŔobotHW`-class in ROS1.
   #. Do not forget the ``PLUGINLIB_EXPORT_CLASS`` macro at the end of the .cpp file.
#. Create .xml library description for the pluginlib, for example see `RRBotSystemPositionOnlyHardware <https://github.com/ros-controls/ros2_control_demos/blob/master/ros2_control_demo_hardware/ros2_control_demo_hardware.xml>`_.

Controller Migration
--------------------
An excellent example of a migrated controller is the `JointTrajectoryController`_.
The real-time critical methods are marked as such.

#. Implement `ControllerInterface`_ class as follows:
   #. If there are any member variables, initialized those in the constructor.
   #. In the `init` method, first call ``ControllerInterface::init`` initialize lifecycle of the controller.
      Then declare all parameters defining their default values.

>    #. Define ``*_interface_configuration`` methods for required command and state interfaces.

Reviewer: ```suggestion
   #. Implement the ``state_interface_configuration()`` and ``command_interface_configuration()`` methods.
```
better to not use `*` here

>    #. Implement ``update`` function for the controller. (**real-time**)

Reviewer: ```suggestion
   #. Implement ``update()`` function for the controller. (**real-time**)
```
can we have `()` after functions? it's an easier notation IMHO

   #. Then implement required lifecycle methods (others are optional):
      * ``on_configure`` - reads parameters and configures controller.

>       * ``on_activate`` - called when contoroller should be activated (started) (**real-time**)

Reviewer: ```suggestion
      * ``on_activate`` - called when controller is activated (started) (**real-time**)
```

>       * ``on_deactivate`` - called when controller should be deactivated (stopped) (**real-time**)

Reviewer: ```suggestion
      * ``on_deactivate`` - called when controller is deactivated (stopped) (**real-time**)
```
